diff -Nur -x '*.orig' -x '*~' system-config-printer-1.4.3+20140219/newprinter.py system-config-printer-1.4.3+20140219.new/newprinter.py
--- system-config-printer-1.4.3+20140219/newprinter.py	2014-03-14 16:59:02.038916968 +0100
+++ system-config-printer-1.4.3+20140219.new/newprinter.py	2014-03-14 17:00:10.078918790 +0100
@@ -1000,10 +1000,12 @@
             # Keep the UI refreshed while we wait for
             # the drivers query to complete.
             (stdout, stderr) = (self.p.stdout, self.p.stderr)
+            done = False
             while self.p.poll() == None:
                 line = stderr.readline ().strip()
                 if (len(line) > 0):
                     if line == "done":
+                        done = True
                         break
                     try:
                         percentage = float(line)
@@ -1015,7 +1017,7 @@
                     Gtk.main_iteration ()
                 if not line:
                     time.sleep (0.1)
-            if self.p.returncode != 0:
+            if self.p.returncode != 0 and not done:
                 ret = False
         except:
             # Problem executing command.
